File: 31-next-permutation/next-permutation.java
class Solution {

    public int[] swap(int [] nums, int start, int end){
        while(start < end){
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
        return nums;
    }

    public void nextPermutation(int[] nums) {
        
        int n = nums.length-1;
        int pivot = -1;
        int index = -1;

        for(int i = n; i>0; i--){
            if(nums[i] > nums[i-1]){
                pivot = nums[i-1];
                index = i-1;
                break;
            }
        }

        if(pivot == -1){
            swap(nums,0,n);
            return;
        }

        for(int i=n; i>0 ; i--){
            if(nums[i] > pivot){
                int temp = nums[i];
                nums[i] = nums[index];
                nums[index] = temp;
                break;
            }
        }

        swap(nums,index+1,n);
    }
}

File: 31-next-permutation/next-permutation.py
class Solution(object):
    def swap(self, nums, start, end):
        while start<end:
            nums[start],nums[end] = nums[end], nums[start]
            start += 1
            end -= 1
        return nums

    def nextPermutation(self, nums):
        pivot = -1
        index = -1
        n= len(nums)-1

        for i in range(n,0,-1):
            if(nums[i]>nums[i-1]):
                pivot = nums[i-1]
                index = i-1
                break
        
        if(pivot == -1):
            self.swap(nums,0,n)
            return
        

        for i in range(n,index,-1):
            if(nums[i] > nums[index]):
                nums[i],nums[index] = nums[index], nums[i]
                break
        
        self.swap(nums, index+1, n)